pkg/pulseserver/pulse_server.go

```diff
@@ -0,0 +1,185 @@
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			return nil
+		case <-pulseTicker.C:
+			logger.V(4).Infof("Updating heartbeat for maestro instance: %s", instance.ID)
+			instance.UpdatedAt = time.Now()
+			_, err = s.instanceDao.UpSert(ctx, instance)
+			if err != nil {
+				// log and ignore the error and continue to tolerate the intermittent issue
+				logger.Error(fmt.Sprintf("Unable to update heartbeat for maestro instance: %s", err.Error()))
+			}
+		case <-checkTicker.C:
+			logger.V(4).Infof("Checking maestro instances liveness and trigger statusresync for dead instances")
+			if err := s.CheckInstances(ctx); err != nil {
+				// log and ignore the error and continue to tolerate the intermittent issue
+				logger.Error(fmt.Sprintf("Unable to check maestro instances: %s", err.Error()))
+			}
+		}
+	}
+}
+
+// CheckInstances checks all maestro instances' liveness and trigger statusresync for dead instances.
+func (s *pulseServerImpl) CheckInstances(ctx context.Context) error {
+	// lock the Instance with a fail-fast advisory lock context.
+	// this allows concurrent processing of many instances by one or more maestro instances exclusively.
+	lockOwnerID, acquired, err := s.lockFactory.NewNonBlockingLock(ctx, "maestro-instances-pulse-check", db.Instances)
+	// Ensure that the transaction related to this lock always end.
+	defer s.lockFactory.Unlock(ctx, lockOwnerID)
+	if err != nil {
+		return fmt.Errorf("error obtaining the event lock: %v", err)
+	}
+	// skip if the lock is not acquired
+	if !acquired {
+		return fmt.Errorf("failed to acquire the lock as another maestro instance is checking instances")
+	}
+	// Instances not pulsing within the last three check intervals are considered as dead.
+	instances, err := s.instanceDao.FindByUpdatedTime(ctx, time.Now().Add(time.Duration(int64(-3*time.Second)*s.checkInterval)))
+	if err != nil {
+		return fmt.Errorf("unable to get outdated maestro instances: %s", err.Error())
+	}
+	deletedInstanceIDs := []string{}
+	for _, i := range instances {
+		deletedInstanceIDs = append(deletedInstanceIDs, i.ID)
+	}
+
+	if len(deletedInstanceIDs) > 0 {
+		// trigger statusresync for dead instances only once even if there are multiple dead instances
+		// will retry in next check if the statusresync fails
+
+		// send resync request to each consumer
+		// TODO: optimize this to only resync resource status for necessary consumers
+		consumerIDs := []string{}
+
+		consumers, err := s.consumerDao.All(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to get all consumers: %s", err.Error())
+		}
+
+		for _, c := range consumers {
+			consumerIDs = append(consumerIDs, c.ID)
+		}
+
+		if err := s.sourceClient.Resync(ctx, consumerIDs); err != nil {
+			return fmt.Errorf("unable to trigger statusresync for maestro instance(s): %s", err.Error())
+		}
+
+		// batch delete dead instances
+		if err := s.instanceDao.DeleteByIDs(ctx, deletedInstanceIDs); err != nil {
```

Comment:
hrm, this is strange, do we really need to delete the instance. Could we just mark the instance as resync is sent?

Reply:
Anyway, we need to persistent the instance status to DB, otherwise, maestro needs to handle two many stale instances(call `OnInstanceDown`).
Marking the instance is resynced in memory is one option, but on maestro restart, the stale instances are still needed to handled.
I prefer to mark the instance deleted in DB.